src/components/Rocket.js

```diff
@@ -0,0 +1,89 @@
+import '../styles/Rocket.css';
+import PropTypes from 'prop-types';
+import { useEffect, useState, useRef } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { buttonReducer } from '../redux/rockets/RocketSlice';
+import getRockets from '../redux/rockets/RocketAPI';
+
+const Rocket = ({
+  id, imgURL, title, description,
+}) => {
+  const [buttonText, setButtonText] = useState('Reserve Rocket');
+  const { reserveState } = useSelector((store) => store.rocket);
+  const buttonRef = useRef(null);
+
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    reserveState.forEach((item) => {
+      if (item.id === id) {
+        setButtonText('Cancel Reservation');
+        buttonRef.current.classList.add('buttonCancel');
+      }
+    });
+  }, [id, reserveState]); // Added 'id' and 'reserveState' to dependency array
+
+  function setButton(e) {
+    if (buttonText === 'Reserve Rocket') {
+      setButtonText('Cancel Reservation');
+      dispatch(buttonReducer({ id, buttonText, title }));
+      e.target.classList.add('buttonCancel');
+    } else {
+      setButtonText('Reserve Rocket');
+      dispatch(buttonReducer({ id, buttonText, title }));
+      e.target.classList.remove('buttonCancel');
+    }
+  }
+
+  return (
+    <div className="rocket-component">
+      <img src={imgURL} alt="rocket" />
+      <span>
+        <h3>{title}</h3>
+        <p>{description}</p>
+        <button
+          ref={buttonRef}
+          onClick={(e) => setButton(e)}
+          className="reserve"
+          type="button"
+        >
+          {buttonText}
+        </button>
+      </span>
+    </div>
+  );
+};
+
+const RocketPage = () => {
+  const dispatch = useDispatch();
+  const { content, info } = useSelector((store) => store.rocket);
+  useEffect(() => {
+    dispatch(getRockets());
+  }, []);
```

Comment:
The problem is whenever you go to rockets page it reFetch all the rockets even when they are available in the state making unnecessary calls to API and burdening the server. 


## HOW TO DO THAT:--

You can achieve the above by checking if rockets are available in REDUX state or not and fetching if their length is zero:-

```
const Rockets = () => {
  const { rockets } = useSelector((store) => store.rockets);
  const dispatch = useDispatch();
  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(fetchRockets());
    }
  }, [dispatch, rockets.length]);
```